Error uninstalling flapjack 1.3.0rc1 - Error generating or dispatching SMS Nexmo message

```
Stopping flapjack
Flapjack stopping...Stopped PID: 143 at /var/run/flapjack/flapjack.pid
 done.
waiting for a max of 30 seconds for process 143 to exit.flapjack[143]: 2015-02-13T07:54:04+00:00 [ERROR] :: flapjack-sms_nexmo :: Error generating or dispatching SMS Nexmo message: TypeError: no implicit conversion of nil into String
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/data/alert.rb:59:in `update'
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/data/alert.rb:59:in `digest'
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/data/alert.rb:59:in `initialize'
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/data/alert.rb:133:in `new'
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/data/alert.rb:133:in `next'
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/gateways/sms_nexmo.rb:48:in `start'
/opt/flapjack/embedded/lib/ruby/gems/2.1.0/gems/flapjack-1.3.0rc1/lib/flapjack/pikelet.rb:126:in `block in start'
flapjack[143]: 2015-02-13T07:54:04+00:00 [ERROR] :: flapjack-processor :: Invalid event data received, Event hash has missing keys 'state', 'entity', 'check' {"type"=>"noop"}
```

--- a/lib/flapjack/data/alert.rb
+++ b/lib/flapjack/data/alert.rb
@@ -130,6 +130,7 @@ module Flapjack
           end
           return nil
         end
+        return if 'shutdown'.eql?(parsed['notification_type'])
         self.new( parsed, :logger => opts[:logger] )
       end
 

--- a/lib/flapjack/data/event.rb
+++ b/lib/flapjack/data/event.rb
@@ -132,12 +132,16 @@ module Flapjack
         elsif options[:archive_events]
           redis.expire(archive_dest, options[:events_archive_maxage])
         end
+        if options[:logger]
+          options[:logger].info parsed.inspect
+        end
         self.new(parsed)
       end
 
       def self.parse_and_validate(raw, opts = {})
         errors = []
         if parsed = ::Flapjack.load_json(raw)
+          return if 'noop'.eql?(parsed['type'])
           if parsed.is_a?(Hash)
             errors = validation_errors_for_hash(parsed, opts)
           else

--- a/lib/flapjack/data/event.rb
+++ b/lib/flapjack/data/event.rb
@@ -132,9 +132,6 @@ module Flapjack
         elsif options[:archive_events]
           redis.expire(archive_dest, options[:events_archive_maxage])
         end
-        if options[:logger]
-          options[:logger].info parsed.inspect
-        end
         self.new(parsed)
       end
 

--- a/lib/flapjack/data/migration.rb
+++ b/lib/flapjack/data/migration.rb
@@ -205,7 +205,7 @@ module Flapjack
 
         return if redis.exists('validated_scheduled_maintenance_periods')
 
-        Flapjack::Data::EntityCheck.all(:redis => redis).select {|ec|
+        Flapjack::Data::EntityCheck.all(:redis => redis).compact.select {|ec|
           ec.in_scheduled_maintenance?
         }.each do |check|
           check.update_current_scheduled_maintenance(:revalidate => true)

--- a/lib/flapjack/data/notification.rb
+++ b/lib/flapjack/data/notification.rb
@@ -85,8 +85,9 @@ module Flapjack
           if options[:logger]
             options[:logger].warn("Error deserialising notification json: #{e}, raw json: #{raw.inspect}")
           end
-          return nil
+          return
         end
+        return if 'shutdown'.eql?(parsed['type'])
         self.new( parsed )
       end
 

--- a/lib/flapjack/gateways/aws_sns.rb
+++ b/lib/flapjack/gateways/aws_sns.rb
@@ -45,7 +45,8 @@ module Flapjack
         until @should_quit
           begin
             @logger.debug("aws_sns gateway is going into blpop mode on #{queue}")
-            deliver( Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger) )
+            alert = Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger)
+            deliver(alert) unless alert.nil?
           rescue => e
             @logger.error "Error generating or dispatching AWS SNS message: #{e.class}: #{e.message}\n" +
               e.backtrace.join("\n")

--- a/lib/flapjack/gateways/email.rb
+++ b/lib/flapjack/gateways/email.rb
@@ -51,7 +51,8 @@ module Flapjack
         until @should_quit
           begin
             @logger.debug("email gateway is going into blpop mode on #{queue}")
-            deliver( Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger) )
+            alert = Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger)
+            deliver(alert) unless alert.nil?
           rescue => e
             @logger.error "Error generating or dispatching email message: #{e.class}: #{e.message}\n" +
               e.backtrace.join("\n")

--- a/lib/flapjack/gateways/sms_messagenet.rb
+++ b/lib/flapjack/gateways/sms_messagenet.rb
@@ -45,7 +45,8 @@ module Flapjack
         until @should_quit
           begin
             @logger.debug("sms_messagenet gateway is going into blpop mode on #{queue}")
-            deliver( Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger) )
+            alert = Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger)
+            deliver(alert) unless alert.nil?
           rescue => e
             @logger.error "Error generating or dispatching SMS Messagenet message: #{e.class}: #{e.message}\n" +
               e.backtrace.join("\n")

--- a/lib/flapjack/gateways/sms_nexmo.rb
+++ b/lib/flapjack/gateways/sms_nexmo.rb
@@ -45,7 +45,8 @@ module Flapjack
         until @should_quit
           begin
             @logger.debug("sms_nexmo gateway is going into blpop mode on #{queue}")
-            deliver( Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger) )
+            alert = Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger)
+            deliver(alert) unless alert.nil?
           rescue => e
             @logger.error "Error generating or dispatching SMS Nexmo message: #{e.class}: #{e.message}\n" +
               e.backtrace.join("\n")

--- a/lib/flapjack/gateways/sms_twilio.rb
+++ b/lib/flapjack/gateways/sms_twilio.rb
@@ -49,7 +49,8 @@ module Flapjack
         until @should_quit
           begin
             @logger.debug("sms_twilio gateway is going into blpop mode on #{queue}")
-            deliver( Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger) )
+            alert = Flapjack::Data::Alert.next(queue, :redis => @redis, :logger => @logger)
+            deliver(alert) unless alert.nil?
           rescue => e
             @logger.error "Error generating or dispatching SMS Twilio message: #{e.class}: #{e.message}\n" +
               e.backtrace.join("\n")

--- a/lib/flapjack/notifier.rb
+++ b/lib/flapjack/notifier.rb
@@ -66,7 +66,7 @@ module Flapjack
         notification = Flapjack::Data::Notification.next(@notifications_queue,
                                                          :redis => @redis,
                                                          :logger => @logger)
-        process_notification(notification) unless notification.nil? || (notification.type == 'shutdown')
+        process_notification(notification) unless notification.nil?
       end
 
       @logger.info("Exiting main loop.")

